File: crawler/nexon/nexon.go
package nexon

import (
	"context"
	"encoding/json"
	"geekermeter-data/crawler"
	"io/ioutil"
	"log"
	"strconv"
	"time"

	"github.com/chromedp/cdproto/cdp"
	"github.com/chromedp/chromedp"
)

var (
	baseURL = "https://career.nexon.com/user/recruit/notice/noticeList"
)

func Nexon() []crawler.Job {
	var crawledData []crawler.Job
	ctx, cancel := chromedp.NewContext(context.Background())
	defer cancel()

	var loc string
	err := chromedp.Run(ctx,
		chromedp.Navigate(baseURL),
		chromedp.Location(&loc),
	)
	crawler.ErrHandler(err)

	err = chromedp.Run(ctx,
		chromedp.Click("#container > ul > li:nth-child(1)", chromedp.NodeVisible))
	crawler.ErrHandler(err)

	var totalPageNode []*cdp.Node
	var totalPage string
	err = chromedp.Run(ctx,
		chromedp.Nodes("#con_right > div.content > div > a.last", &totalPageNode))
	crawler.ErrHandler(err)

	for _, row := range totalPageNode {
		totalPage = row.AttributeValue("href")[20:22]
	}
	t, _ := strconv.Atoi(totalPage)

	for i := 0; i <= t; i++ {
		var nodes []*cdp.Node
		var titleNode []*cdp.Node
		var newbieNode []*cdp.Node
		var countNode []*cdp.Node

		err := chromedp.Run(ctx,
			chromedp.Sleep(2*time.Second),
			chromedp.Nodes("#con_right > div.content > table > tbody > tr > td.tleft.fc_02 > a", &nodes, chromedp.ByQueryAll),
			chromedp.Nodes("#con_right > div.content > table > tbody > tr > td.tleft.fc_02 > a > span", &titleNode, chromedp.ByQueryAll),
			chromedp.Nodes("#con_right > div.content > table > tbody > tr > td:nth-child(2)", &newbieNode, chromedp.ByQueryAll),
			chromedp.Nodes("#con_right > div.content > table > tbody", &countNode, chromedp.ByQueryAll),
		)
		crawler.ErrHandler(err)
		var count int64
		for _, row := range countNode {
			count = row.ChildNodeCount
		}
		temp := make([]crawler.Job, count)

		for l, row := range nodes {
			temp[l].URL = "https://career.nexon.com" + row.AttributeValue("href")
		}
		for l, row := range titleNode {
			temp[l].Title = row.Children[0].NodeValue
			temp[l].Origin = "nexon"
		}
		for l, row := range newbieNode {
			temp[l].Newbie = row.Children[0].NodeValue
		}

		for i := 1; i <= int(count); i++ {
			t := strconv.Itoa(i)
			err := chromedp.Run(ctx,
				chromedp.Text("#con_right > div.content > table > tbody > tr:nth-child("+t+") > td:nth-child(6)", &temp[i-1].StartDate),
			)
			crawler.ErrHandler(err)
		}

		crawledData = append(crawledData, temp...)

		if i != t {
			err = chromedp.Run(ctx,
				chromedp.Click("#con_right > div.content > div > a.next", chromedp.NodeVisible),
			)
			crawler.ErrHandler(err)
		}
	}
	return crawledData
}

func BodyText(box crawler.Job) { //현재 쓸데없는 값까지 하는 중->예외처리 실패로 인해..
	ctx, cancel := chromedp.NewContext(context.Background())
	defer cancel()

	// run task list
	var loc string
	err := chromedp.Run(ctx,
		chromedp.Navigate(box.URL),
		chromedp.Location(&loc),
	)
	crawler.ErrHandler(err)
	box.Content = make([]string, 1)
	err = chromedp.Run(ctx,
		chromedp.Text("#con_right > div.content > div.list_txt01",
			&box.Content[0]),
	)
	log.Println(box.Content[0])
	crawler.ErrHandler(err)
	doc, _ := json.Marshal(box)
	_ = ioutil.WriteFile("./dataset/tmp/"+crawler.Exceptspecial(box.URL)+".json", doc, 0644)

}

func Start() {
	log.Println("Start crawl Nexon")
	list := Nexon()
	log.Println("End crawl Nexon")
	for _, row := range list {
		BodyText(row)
	}
}

File: crawler/nexon/nexon.py
import re, os, sys
import selenium
sys.path.append(os.path.dirname(os.path.abspath(os.path.dirname(__file__))))
from util import *
from selenium import webdriver
from bs4 import BeautifulSoup
import warnings
warnings.filterwarnings(action='ignore')


def nexon(driver_path=None):
    if not driver_path:
        driver_path='/Users/mingihong/chromedriver'
    driver = headless(driver_path)
#    driver = webdriver.Chrome(driver_path)
    driver.get('https://career.nexon.com/user/recruit/notice/noticeList')
    WebDriverWait(driver, 10).until(
        EC.presence_of_element_located((By.CSS_SELECTOR, "#container > ul > li:nth-child(1) > a > img"))
    )
    driver.find_element_by_css_selector('#container > ul > li:nth-child(1) > a > img').click()
    final_data = []
    while True:
        crawled_data = []
        WebDriverWait(driver, 10).until(
            EC.presence_of_element_located((By.ID, "con_right"))
        )
        html = driver.page_source
        soup = BeautifulSoup(html,'html.parser')

        posts = soup.select('#con_right > div.content > table > tbody > tr')
        for post in posts:
            post_date = transfrom_date(post.select('td')[5].text)
            if not post_date:
                continue
            post_title = post.select('td.tleft.fc_02 > a > span')[0].text
            post_url = 'https://career.nexon.com'+post.select('td.tleft.fc_02 > a')[0].get('href')
            post_newbie = post.select('td')[1].text

            crawled_data.append(
                job(post_title,post_url,'nexon',post_date,post_newbie).data
            )

        final_data.extend(crawled_data)

        try:
            driver.find_element_by_css_selector('#con_right > div.content > div > a.next').click()
        except:
            break


    return final_data, driver

def body_text(driver,json):
    driver.get(json['url'])

    html = driver.page_source
    soup = BeautifulSoup(html,'html.parser')
    txt = re.sub('[\s]+',' ',soup.select('#con_right > div.content > div.list_txt01')[0].text)
    json['contents'] += txt
    return json

def run(driver_path=None):
    print('start crawling : nexon')
    json_list, driver = nexon(driver_path)
    print('The number of nexon post : %d'%(len(json_list)))
    for i,json in enumerate(json_list):
        if i%10 == 0:
            print('nexon post : %d'%(i))
        jojsonb = body_text(driver,json)
    driver.quit()
    print('finish crawling : nexon')

    return json_list


if __name__ == "__main__":
    run()
